rust/kernel/rbtree.rs

```diff
@@ -418,6 +571,210 @@ impl<K, V> Drop for RBTree<K, V> {
     }
 }
 
+/// Direction for `RBTreeCursor` operations
+pub enum Direction {
+    /// the node immediately before, in sort order
+    Prev,
+    /// the node immediately after, in sort order
+    Next,
+}
+
+/// A bidirectional cursor over the tree nodes, sorted by key
+pub struct RBTreeCursor<'a, K, V> {
+    _tree: PhantomData<&'a RBTree<K, V>>,
+    root: bindings::rb_root,
+    current: *mut bindings::rb_node,
+}
```

Comment:
I don't know how the Linux rbtree works, but I'm concerned that you're making a copy of the root object. How about storing a `*mut rb_root` instead?

Reply:
I added this when implementing the remove function (only place it's used), because it's required by rb_erase. IIUC all I need is a pointer there anyway so that should be fine.